spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/RemoveCarryoverIterator.java

```diff
@@ -0,0 +1,153 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark;
+
+import java.util.Iterator;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.types.StructType;
+
+/**
+ * An iterator that removes the carry-over rows from changelog tables within a single Spark task. It
+ * assumes that rows are partitioned by identifier(or all) columns, and it is sorted by both
+ * identifier(or all) columns and change type.
+ *
+ * <p>Carry-over rows are the result of a removal and insertion of the same row within an operation
+ * because of the copy-on-write mechanism. For example, given a file which contains row1 (id=1,
+ * data='a') and row2 (id=2, data='b'). A copy-on-write delete of row2 would require erasing this
+ * file and preserving row1 in a new file. The change-log table would report this as follows,
+ * despite it not being an actual change to the table.
+ *
+ * <ul>
+ *   <li>(id=1, data='a', op='DELETE')
+ *   <li>(id=1, data='a', op='INSERT')
+ *   <li>(id=2, data='b', op='DELETE')
+ * </ul>
+ *
+ * The iterator finds the carry-over rows and removes them from the result. For example, the above
+ * rows will be converted to:
+ *
+ * <ul>
+ *   <li>(id=2, data='b', op='DELETE')
+ * </ul>
+ */
+class RemoveCarryoverIterator extends ChangelogIterator {
+  private final int[] indicesToIdentifySameRow;
+
+  private Row deletedRow = null;
+  private long deletedRowCount = 0;
+  private Row cachedNextRecord = null;
+
+  RemoveCarryoverIterator(Iterator<Row> rowIterator, StructType rowType) {
+    super(rowIterator, rowType);
+    this.indicesToIdentifySameRow = generateIndicesToIdentifySameRow(rowType.size());
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (hasCachedDeleteRow() || cachedNextRecord != null) {
+      return true;
+    }
+    return rowIterator().hasNext();
+  }
+
+  @Override
+  public Row next() {
+    if (returnCachedDeleteRow()) {
+      deletedRowCount--;
+      return deletedRow;
+    }
+
+    Row currentRow = currentRow();
+
+    if (currentRow.getString(changeTypeIndex()).equals(DELETE) && rowIterator().hasNext()) {
+      // cache the delete row if there is 0 delete row cached
+      if (!hasCachedDeleteRow()) {
+        deletedRow = currentRow;
+        deletedRowCount++;
+      }
+
+      Row nextRow = rowIterator().next();
+
+      if (isSameRecord(currentRow, nextRow)) {
+        if (nextRow.getString(changeTypeIndex()).equals(INSERT)) {
+          deletedRowCount--;
+        } else {
+          deletedRowCount++;
+        }
+      } else {
+        // mark the boundary since the next row is not the same record as the current row
+        cachedNextRecord = nextRow;
```

Comment:
I'm not sure this behavior is correct, my gut says there needs to be a while loop in here

Reply:
Rewrote it with while-loop in the new commit.